### ISSUE TITLE
Weird values on heat index calculations

### ISSUE THREAD

**Issue by RobertDB59** [RobTillaart/Temperature#10]
When using the formula's given in the library temperature.h the value's I am getting are way off to the values when I use an online heat index calculator. This is for both formula's °C and °F. When I use the following formula: 

`HI = 0.5 * (T + 61.0 + ((T-68.0)*1.2) + (RH*0.094)) `

found at https://www.wpc.ncep.noaa.gov/html/heatindex_equation.shtml it returns imho a more correct value.

Maybe I am missing something?

**Comment by RobTillaart** [comment]
A quick look shows I only implemented the first formula without the adjustments.

From the link
_In practice, the simple formula is computed first and the result averaged with the temperature. If this heat index value is 80 degrees F or higher, the full regression equation along with any adjustment as described above is applied._

So code needs to be reworked.

Note: **HeatIndexC()** should become a wrapper around **HeatIndex()** to minimize footprint and reduce chance on errors.

**Comment by RobTillaart** [comment]
I got my formulas from - https://en.wikipedia.org/wiki/Heat_index which does not mention the NOAA adjustments.
As NOAA is the authority I follow them. 
(found a small can of time, so I start on it now)

**Comment by RobTillaart** [comment]
@RobertDB59 
Created develop branch - https://github.com/RobTillaart/Temperature/tree/develop + PR

Can you verify if this works as expected?

**Comment by RobTillaart** [comment]
@RobertDB59 

> Can you verify if this works as expected?

Checked the table generated with the example with the table at https://www.calculator.net/heat-index-calculator.html
and they matched quite well except the column for 80.  Fixed a >  to a >=  and the 80 column matched too.

Will push the change today an merge the changes. 

Thanks again for pointing out the issue,
Rob

**Comment by RobertDB59** [comment]
Hi Rob,
Just verified the code and is working as expected. 

Keep up the good work.
Robert

**Comment by RobertDB59** [comment]
I think there is a minor typo 

` HI += ((13 - RH) / 4) * sqrt((17 - abs(TF - 95.0)) / 17);`

Shouldn't this be subtracted instead of added to HI?

**Comment by RobTillaart** [comment]
> I think there is a minor typo
> 
> ` HI += ((13 - RH) / 4) * sqrt((17 - abs(TF - 95.0)) / 17);`
> 
> Shouldn't this be subtracted instead of added to HI?

Checked the website again and it is a **major** typo

_where T is temperature in degrees F and RH is relative humidity in percent.  HI is the heat index expressed as an apparent temperature in degrees F.  If the RH is less than 13% and the temperature is between 80 and 112 degrees F, then the following adjustment is **subtracted** from HI:_

```ADJUSTMENT = [(13-RH)/4]*SQRT{[17-ABS(T-95.)]/17}```

_where ABS and SQRT are the absolute value and square root functions, respectively.  On the other hand, if the RH is greater than 85% and the temperature is between 80 and 87 degrees F, then the following adjustment is **added** to HI:_

```ADJUSTMENT = [(RH-85)/10] * [(87-T)/5]```

A 0.3.1 release is coming soon ...

**Comment by RobTillaart** [comment]
updated, merged and released 0.3.1
added a few unit tests
please verify

**Comment by RobertDB59** [comment]
Verified and it seems to be all good. Just an idea, maybe you could use the celsius to fahrenheid calculation and vice versa already present in your library for the heat index calculation in a future release. Thanks again en een fijne dag verder.

**Comment by RobTillaart** [comment]
> maybe you could use the celsius to fahrenheid calculation and vice versa already present in your library for the heat index calculation in a future release. 

Already did that, but I left the code with all the constants in comments. I might consider a Celsius only heatindex once, that would safe some float math. On the other hand the gain would be most 10% as the float math is already pretty heavy.

Check in code

```cpp
float heatIndexC(float celcius, float humidity)
{
  float TF = celcius * (9.0 / 5.0) + 32;
  float HI = (heatIndex(TF, humidity) - 32) * (5.0 / 9.0);
  return HI;
...
```

>  en een fijne dag verder.

Dank je wel !  
(you speak Dutch?  I live in East part of Brabant, near Helmond)

**Comment by RobertDB59** [comment]
Born and raised in Eindhoven, close to Helmond, and I moved 12 years ago to the south of France. So yes I speak Dutch as well.

**Comment by RobertDB59** [comment]
What I was trying to say is that you could use 

`float TF = Fahrenheid(celcius);`

instead of 

`TF = celcius * (9.0 / 5.0) + 32;`

Btw celcius is written with an S in the middle as celsius. Maybe you have done that on purpose.

**Comment by RobTillaart** [comment]
My daughter lives near the High Tech Campus, and I studied there so Eindhoven is very familiar.
 
Both Celcius and Celsius exist, see on this page - https://en.wikipedia.org/wiki/Talk%3ACelsius
But I agree as the man was called Celsius that should be it. 

I will update it soon as I also wanted to add the Rankine and Reamure temperature scale conversions too.
(need to make an issue for that)

I close this issue, as the problem is solved. If there are other issues, please let me know.

**Comment by RobTillaart** [comment]
@RobertDB59 

Bad weather over here so I am working on a update - https://github.com/RobTillaart/Temperature/pull/14
with a convertor class.

**Comment by RobertDB59** [comment]
Hallo Rob,

Misschien kun je de volgende snippet gebruiken voor de luchtdruk berekening
naar zeeniveau.
Wij wonen op 430 meter hoogte waardoor de de hier heersende luchtdruk niet
overeenkomt met een meetstation in Perpignan.

// calculate the barometric pressure at sea level
float baroReadingC( float pressure, float celsius, float altitude){
  return pressure * pow( 1 - (( 0.0065 * altitude) / ( celsius + 0.0065 *
altitude + 273.15)), -5.257);
}

Groet, Robert

Op za 8 jan. 2022 om 16:02 schreef Rob Tillaart ***@***.***>:

**Comment by RobTillaart** [comment]
Thank you for this function, I will add it to the temperature library.
My proposal is to change the name as that is more descriptive.

```cpp
float baroToSeaLevelC( float pressure, float celsius, float altitude)
{
  float altitudeFactor = 0.0065 * altitude;
  float kelvin = celsius + 273.15;
  return pressure * pow( 1 - (altitudeFactor / (kelvin + altitudeFactor)), -5.257);
}
```

Q: Do you have a reference source (NOAA or so) for the code?

found this 
- https://www.engineeringtoolbox.com/air-altitude-pressure-d_462.html
but it does not have the temperature correction ( but it has the -5.257 exponent)
- https://www.omnicalculator.com/physics/air-pressure-at-altitude#how-to-calculate-air-pressure-at-altitude
- https://en.wikipedia.org/wiki/Barometric_formula
Different but I assume if I do the math it will work.

update: added the function in upcoming 0.3.2 release

**Comment by RobertDB59** [comment]
It is called the Hypsometric equation, also known as the thickness
equation, relates an atmospheric pressure ratio to the equivalent thickness
of an atmospheric layer. There are multiple sources on the web explaining
this calculation. I don't remember what authority I used as reference.

Op zo 9 jan. 2022 om 10:21 schreef Rob Tillaart ***@***.***>:

**Comment by RobTillaart** [comment]
OK, found some references and they match the equation (enough). So merged into the library.
thanks again.

**Comment by RobertDB59** [comment]
I am not very familiar with github to propose modifications or add stuff so
I keep using this thread for now.

Maybe it is adviced to change the "wind chill" function.
As https://en.wikipedia.org/wiki/Wind_chill states " Windchill temperature
is defined only for temperatures at or below 10 °C (50 °F) and wind speeds
above 4.8 kilometres per hour (3.0 mph).[11]
<https://en.wikipedia.org/wiki/Wind_chill#cite_note-noaa-11>
I added following line to my function ( I have evertything metric :) )

if( kilometerPerHour < 4.8 || celsius > 10) wchill = celsius;
return wchill;

Same as for heat index,  temperature equal to or greater than 80 °F (27 °C)
and relative humidity equal to or greater than 40%, also from the wiki page.
Otherwise heat index equals air temperature.

Hope this helps

Op zo 9 jan. 2022 om 11:50 schreef Rob Tillaart ***@***.***>:

**Comment by RobTillaart** [comment]
Hi Robert,

> I am not very familiar with github to propose modifications or add stuff so
I keep using this thread for now.

You can always create a new issue if it concerns a new topic, or reopen an exiting one if related. 

These are good remarks, I will add them for a new 0.3.3 release. 
This is the way open source grows!

**Comment by RobTillaart** [comment]
https://www.wpc.ncep.noaa.gov/html/heatindex_equation.shtml has some other definition about below 80F range.

I reread it 

_The Rothfusz regression is not appropriate when conditions of temperature and humidity warrant a heat index value below about 80 degrees F. In those cases, a simpler formula is applied to calculate values consistent with Steadman's results:_

**HI = 0.5 * {T + 61.0 + [(T-68.0)*1.2] + (RH*0.094)}**

_In practice, **the simple formula is computed first** and the result averaged with the temperature. If this heat index value is 80 degrees F or higher, the full regression equation along with any adjustment as described above is applied._

and I notice I do not compute this simple first so I have to fix this.
(good reading is so difficult, mea culpa)

**Comment by RobTillaart** [comment]
Changes implemented in develop branch and PR created -  https://github.com/RobTillaart/Temperature/pull/15

update: After some failures the unit test run successful again.

**Comment by RobTillaart** [comment]
Merged PR and released 0.3.3 version.
